_posts/2015-12-14-simply-secure-rails-api.markdown

```diff
@@ -0,0 +1,733 @@
+> 
+> Working at a company like WeWork, I (We) believe that individual success is exponentially more achievable when you're part of a greater community. 
+> 
+> This belief can sometimes mean uncomfortable travels in which trust outweighs mistrust and you are willing to take on risk in the interest of greater success for both yourself and your community.
+> 
+> Ok. Now that's out of the way. What follows is a system/explanation, undoubtedly flawed, but one which We hope will help some folks to understand some security concepts a little better and maybe help them in there travels. 
+> 
+> Thank you, and as always, comments and feedback are not only welcome, but necessary.
+
+
+
+---
+
+
+
+If you're attempting to implement a service orientated architecture 
+one of the most daunting tasks can be deciding on a method for 
+securing API endpoints.
+
+A simple (and obvious) solution is to make every request using an 
+API key or something like [Basic HTTP Authentication headers](https://en.wikipedia.org/wiki/Basic_access_authentication).
+
+Something like:
+
+Requesting APP (using the wonderful [Faraday](https://github.com/lostisland/faraday)):
+
+```ruby
+  Faraday.new(:url => "https://www.api.io/api/v1/posts") do |faraday|
+    faraday.adapter Faraday.default_adapter
+    faraday.use Faraday::Request::TokenAuthentication,
+    faraday.headers['Content-Type'] = 'application/json'
+    faraday.token_auth("1234567cy")
+  end
+```
+
+Receiving APP:
+
+```ruby
+  module Api
+    module V1
+      class PostsController < ApplicationController
+        before_filter :restrict_access
+
+        def index
+          @posts = Post.page(1)
+          respond_with @posts
+        end
+
+        private
+
+        def restrict_access        
+          authenticate_or_request_with_http_token do |token, options|          
+            ApiKey.exists?(access_token: token)
+          end
+        end
+
+      end
+    end
+  end
+```
+
+This is a fine system, pleasingly simple and straight-forward.
+However, it misses one basic fact about these requests: Most 
+are made on behalf of the current **user and not the requesting
+application.** 
+
+So a posts controller:
+
+```ruby
+  module Api
+    module V1
+      class PostsController < ApplicationController
+        before_filter :restrict_access
+
+        def index
+          @posts = current_user.posts.page(1)
+          respond_with @posts
+        end
+
+        private
+
+        def restrict_access        
+          authenticate_or_request_with_http_token do |token, options|          
+            ApiKey.exists?(access_token: token)
+          end
+        end
+
+      end
+    end
+  end
+```
+
+Is, well, a little awkward. Because there's really no way to 
+know who the "current_user" is without passing in some extra variable.
```

Comment:
Re our discussion in person -- I think our real motivation in using this system is so that we can give out these tokens to the *clients* which are untrusted applications. For trusted applications the plain uuid + the master token is OK (and that's what we do between our services in many cases). But we don't want clients to be able to forge requests for other users. Explain that use case.